File: backend/internal/model/websocket.go
package model

// WebSocketMessageType WebSocket message type constants
const (
	MessageTypeCodeUpdate = "code_update"
	MessageTypeGameStart  = "game_start"
	MessageTypeGameEnd    = "game_end"
)

// SubmissionStatusMessage submission status message
type SubmissionStatusMessage struct {
	Type            string   `json:"type"`
	MatchID         string   `json:"match_id"`
	UserID          string   `json:"user_id"`
	Status          string   `json:"status"` // "started", "completed", "failed"
	Passed          *bool    `json:"passed,omitempty"`
	TotalTestCases  *int     `json:"total_test_cases,omitempty"`
	PassedTestCases *int     `json:"passed_test_cases,omitempty"`
	ExecutionTime   *float64 `json:"execution_time,omitempty"`
	MemoryUsage     *float64 `json:"memory_usage,omitempty"`
	Message         string   `json:"message"`
	Timestamp       int64    `json:"timestamp"`
}

// TestCaseDetailMessage test case detail information message
type TestCaseDetailMessage struct {
	Type           string      `json:"type"`
	MatchID        string      `json:"match_id"`
	UserID         string      `json:"user_id"`
	TestCaseIndex  int         `json:"test_case_index"`
	TotalTestCases int         `json:"total_test_cases"`
	Status         string      `json:"status"`                  // "running", "completed"
	Input          interface{} `json:"input"`                   // Actual input value
	ExpectedOutput interface{} `json:"expected_output"`         // Expected output value
	ActualOutput   interface{} `json:"actual_output,omitempty"` // Actual output value
	Passed         *bool       `json:"passed,omitempty"`
	ExecutionTime  *float64    `json:"execution_time,omitempty"`
	MemoryUsage    *float64    `json:"memory_usage,omitempty"`
	Timestamp      int64       `json:"timestamp"`
}

// CodeUpdateMessage code update message
type CodeUpdateMessage struct {
	Type    string `json:"type"`
	MatchID string `json:"match_id"`
	UserID  string `json:"user_id"`
	Code    string `json:"code"`
}

// MatchEndMessage 매치 종료 메시지
type MatchEndMessage struct {
	Type     string `json:"type"`
	MatchID  string `json:"match_id"`
	WinnerID string `json:"winner_id"`
}

File: frontend/src/types/websocket.ts
// WebSocket 메시지 타입 정의

// TestCase 입출력을 위한 타입 (JSON으로 직렬화 가능한 모든 값)
export type TestCaseValue = string | number | boolean | null | TestCaseValue[] | { [key: string]: TestCaseValue };

export interface SubmissionStatusMessage {
  type: 'submission_started' | 'submission_completed' | 'submission_failed';
  match_id: string;
  user_id: string;
  status: 'started' | 'completed' | 'failed';
  passed?: boolean;
  total_test_cases?: number;
  passed_test_cases?: number;
  execution_time?: number;
  memory_usage?: number;
  message: string;
  timestamp: number;
}

export interface TestCaseDetailMessage {
  type: 'test_case_running' | 'test_case_completed';
  match_id: string;
  user_id: string;
  test_case_index: number;
  total_test_cases: number;
  status: 'running' | 'completed';
  input: TestCaseValue;
  expected_output?: TestCaseValue;
  expected?: TestCaseValue; // Backend sends this field
  actual_output?: TestCaseValue;
  actual?: TestCaseValue; // Backend sends this field
  passed?: boolean;
  execution_time?: number;
  memory_usage?: number;
  timestamp: number;
}

export interface TestCaseResult {
  index: number;
  input: TestCaseValue;
  expectedOutput: TestCaseValue;
  actualOutput?: TestCaseValue;
  passed?: boolean;
  status: 'pending' | 'running' | 'completed';
  executionTime?: number;
  memoryUsage?: number;
}

export interface SubmissionProgress {
  isSubmitting: boolean;
  totalTestCases: number;
  completedTestCases: number;
  testCaseResults: TestCaseResult[];
  overallPassed?: boolean;
  executionTime?: number;
  memoryUsage?: number;
  statusMessage?: string;
}
